README.md
# Terminal Site

- [Terminal Site](#terminal-site)
  - [Description](#description)
  - [About gh 0.0.11 Update](#about-gh-0011-update)
  - [Tree of the project](#tree-of-the-project)
  - [About the JS](#about-the-js)
  - [About the CSS](#about-the-css)
  - [Updates](#updates)
  - [License](#license)


## Description

We're looking to create a simple, single-page web application that mimics the look and feel of a terminal, and allows users to interact with it by entering commands.

For the lawet, we can use CSS to style the HTML elements to make it look like a terminal. we can use a monospace font and set the background color to black.

To create the command prompt, we can use a simple HTML input element, and use JavaScript to listen for the "enter" key press event.

To read the content of the about.md file, we can use JavaScript's fetch API or XMLHttpRequest to read the file as a text file, and then use the response to display the content in the browser.

To handle the different commands, we can create a JavaScript object that maps commands to functions, and then use the input value to determine which function to call.

To display the output of the command, we can create a new HTML element, such as a div, and use JavaScript to set its innerHTML to the output of the command.

To emulate the terminal experience we can use libraries like xterm.js which can handle terminal emulation in web applications.

The index.html file is the main file of a website, it's where the HTML code for the website is located. It starts with a doctype declaration, which is used to indicate that the document is an HTML document. Then it has a head section which includes the title of the website, a link to the styles.css file, a link to a Google font, and a meta tag for the viewport. The body section includes the structure of the website and includes a div with an id of "terminal" which contains the elements of the website such as the heading, output, prompt and input elements. The script src is also included at the bottom of the file linking the main.js file.

The main.js file is a JavaScript file that is used to add interactivity to the website. It starts by grabbing elements from the DOM (Document Object Model) using the getElementById method, and adding an event listener to the input element that listens for the enter key. When the enter key is pressed, it calls the handleCommand function and passes the input value as an argument. The handleCommand function checks the command value and depending on the command, it updates the output element's innerHTML with the corresponding text.

The styles.css file is a file that is used to style the website. It includes CSS code that styles the elements in the website such as the body, output, terminal, and input elements. It uses CSS selectors to target specific elements and apply styles to them. It also includes media queries which are used to change the styles of elements based on the screen size.


## About gh 0.0.11 Update

In this version, the commands and the content of the about page are stored remotely, so you can add more commands and content without having to update the code. This also improves the security of the website, since the content is not stored in the source code. Performance is also improved, since the content is loaded only when the user requests it.


## Tree of the project

```
* Terminal Site
├── CHANGELOG.md
├── LICENSE.md
├── README.md
├── index.html
├── main.js
└── styles.css
```


## About the JS

```js
const input = document.getElementById("input");
const output = document.getElementById("output");

input.addEventListener("keyup", function (event) {
    if (event.keyCode === 13) {
        event.preventDefault();
        const command = input.value.toLowerCase();
        input.value = "";
        handleCommand(command);
    }
});

async function handleCommand(command) {
    try {
        const response = await fetch('https://raw.githubusercontent.com/superdatas/terminal-text/main/commands.json');
        const commands = await response.json();
        if (commands.commands[command]) {
            commands.commands[command].text.forEach(function (text) {
                output.innerHTML += `<p>${text}</p>`;
            });
        } else {
            output.innerHTML = `<p>Command not found: ${command}</p>`;
        }
    } catch (error) {
        console.error(error);
        output.innerHTML = '<p>An error occurred while loading commands</p>';
    }
}
```


## About the CSS

```css
#output,
#terminal,
body {
    background-color: #000;
    color: #fff
}

body {
    font-family: "Inconsolata", monospace
}

h1 {
    font-size: 2em;
    text-align: center
}

#terminal {
    width: 80%;
    margin: 50px auto;
    padding: 20px;
    border-radius: 10px
}

#output {
    margin-bottom: 20px;
    overflow: auto;
    max-height: 300px
}

#prompt {
    display: flex;
    align-items: center
}

#path {
    font-size: 14px
}

#input {
    width: 100%;
    padding: 10px;
    border: 0;
    border-radius: 10px;
    background-color: #000;
    color: #fff;
    font-family: "Inconsolata", monospace
}

@media only screen and (min-width:375px) {
    #terminal {
        font-size: 1.2em
    }
}

@media only screen and (min-width:414px) {
    #terminal {
        font-size: 1.4em
    }
}

@media only screen and (min-width:390px) {
    #terminal {
        font-size: 1.4em
    }
}

@media only screen and (min-width:428px) {
    #terminal {
        font-size: 1.6em
    }
}

@media only screen and (min-width:810px) {
    #terminal {
        font-size: 1.8em
    }
}

@media only screen and (min-width:384px) {
    #terminal {
        font-size: 1.6em
    }
}

@media only screen and (min-width:412px) {
    #terminal {
        font-size: 1.8em
    }
}

@media only screen and (min-width:30px) {
    #terminal {
        font-size: 1.8em
    }
}
```


## Updates

we can see the updates of the project in the [CHANGELOG.md](CHANGELOG.md) file.


## License

This project is licensed under the Pylar AI creative ML License - see the [LICENSE.md](LICENSE.md) file for details

CHANGELOG.md
# Update Log


- [Update Log](#update-log)
  - [gh 0.0.17 | CSS Optimized](#gh-0017--css-optimized)
  - [gh 0.0.16 | Cool background effect](#gh-0016--cool-background-effect)
  - [gh 0.0.15 | Cool title effect](#gh-0015--cool-title-effect)
  - [gh 0.0.14 | Social links redirects](#gh-0014--social-links-redirects)
  - [gh 0.0.13 | Improved experience on phones](#gh-0013--improved-experience-on-phones)
  - [gh 0.0.12 | Console cleans after last command](#gh-0012--console-cleans-after-last-command)
  - [gh 0.0.11 | Commands and content remotely stored](#gh-0011--commands-and-content-remotely-stored)
  - [gh 0.0.10 | Case sensitive detection](#gh-0010--case-sensitive-detection)
  - [gh 0.0.9 | JS and HTML Content](#gh-009--js-and-html-content)
  - [gh 0.0.8 | CSS Optimized](#gh-008--css-optimized)
  - [gh 0.0.8 | Readme Optimized](#gh-008--readme-optimized)
  - [gh 0.0.7 | Blinked not blinking](#gh-007--blinked-not-blinking)
  - [gh 0.0.6 | CSS Optimized](#gh-006--css-optimized)
  - [gh 0.0.5 | The Viewport Update](#gh-005--the-viewport-update)
  - [gh 0.0.4 | Minimal files](#gh-004--minimal-files)
  - [gh 0.0.3 | It prints about!](#gh-003--it-prints-about)
  - [gh 0.0.1 | Surce Code](#gh-001--surce-code)


## gh 0.0.17 | CSS Optimized
## gh 0.0.16 | Cool background effect
## gh 0.0.15 | Cool title effect
## gh 0.0.14 | Social links redirects
## gh 0.0.13 | Improved experience on phones
## gh 0.0.12 | Console cleans after last command
## gh 0.0.11 | Commands and content remotely stored
## gh 0.0.10 | Case sensitive detection
## gh 0.0.9 | JS and HTML Content
## gh 0.0.8 | CSS Optimized
## gh 0.0.8 | Readme Optimized
## gh 0.0.7 | Blinked not blinking
## gh 0.0.6 | CSS Optimized
## gh 0.0.5 | The Viewport Update

iphone 11 viewport is 375, 11 pro is 414, iPhone 12 mini is 390, iphone 12 is 428, ipad 810, galaxy s20+ is 384, s20 ultra is 412, S20 is 30, adapt the site and fonts to display better on this resolutions

## gh 0.0.4 | Minimal files
## gh 0.0.3 | It prints about!
## gh 0.0.1 | Surce Code

matrix.js
var c = document.getElementById("matrix");
var ctx = c.getContext("2d");
var symbols = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
c.height = window.innerHeight;
c.width = window.innerWidth;

var font_size = 10;
var columns = c.width / font_size;
var drops = [];
for (var x = 0; x < columns; x++) {
    drops[x] = 1;
}

function draw() {
    ctx.fillStyle = "rgba(0, 0, 0, 0.05)";
    ctx.fillRect(0, 0, c.width, c.height);

    ctx.fillStyle = "#0F0";
    ctx.font = font_size + "px arial";
    for (var i = 0; i < drops.length; i++) {
        var text = symbols[Math.floor(Math.random() * symbols.length)];
        ctx.fillText(text, i * font_size, drops[i] * font_size);
        if (drops[i] * font_size > c.height && Math.random() > 0.975) {
            drops[i] = 0;
        }
        drops[i]++;
    }
}

setInterval(draw, 33);


main.js
const input = document.getElementById("input");
const output = document.getElementById("output");

var c = document.getElementById("matrix");
var ctx = c.getContext("2d");
var symbols = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
c.height = window.innerHeight;
c.width = window.innerWidth;

var font_size = 10;
var columns = c.width / font_size;
var drops = [];
for (var x = 0; x < columns; x++) {
    drops[x] = 1;
}

function draw() {
    ctx.fillStyle = "rgba(0, 0, 0, 0.05)";
    ctx.fillRect(0, 0, c.width, c.height);

    ctx.fillStyle = "#0F0";
    ctx.font = font_size + "px arial";
    for (var i = 0; i < drops.length; i++) {
        var text = symbols[Math.floor(Math.random() * symbols.length)];
        ctx.fillText(text, i * font_size, drops[i] * font_size);
        if (drops[i] * font_size > c.height && Math.random() > 0.975) {
            drops[i] = 0;
        }
        drops[i]++;
    }
}

setInterval(draw, 33);

let intervalId;

window.onload = function () {
    matrixTextEffect();
}

function matrixTextEffect() {
    let h1 = document.querySelector("h1");
    let possibleChars = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
    let text = h1.textContent;
    intervalId = setInterval(() => {
        let newText = "";
        for (let i = 0; i < text.length; i++) {
            newText += possibleChars.charAt(Math.floor(Math.random() * possibleChars.length));
        }
        h1.textContent = newText;
    }, 50);

    setTimeout(() => {
        clearInterval(intervalId);
        h1.textContent = "Miguel Gargallo's Terminal";
    }, 3000);
}


input.addEventListener("keyup", function (event) {
    if (event.keyCode === 13) {
        event.preventDefault();
        const command = input.value.toLowerCase();
        input.value = "";
        handleCommand(command);
    }
});

async function handleCommand(input) {
    output.innerHTML = "";
    try {
        const commandsResponse = await fetch('https://raw.githubusercontent.com/superdatas/terminal-text/main/commands.json');
        const commands = await commandsResponse.json();
        if (commands.commands[input]) {
            commands.commands[input].text.forEach(function (text) {
                output.innerHTML += `<p>${text}</p>`;
            });
        } else {
            const actionsResponse = await fetch('https://raw.githubusercontent.com/superdatas/terminal-text/main/actions.json');
            const actions = await actionsResponse.json();
            if (actions.actions[input]) {
                if (/Android|webOS|iPhone|iPad|iPod|BlackBerry|IEMobile|Opera Mini/i.test(navigator.userAgent)) {
                    window.location.replace(actions.actions[input].link[0]);
                } else {
                    window.open(actions.actions[input].link[0]);
                }
            } else {
                output.innerHTML = `<p>Command or action not found: ${input}</p>`;
            }
        }
    } catch (error) {
        console.error(error);
        output.innerHTML = '<p>An error occurred while loading commands and actions</p>';
    }
}
